### ISSUE TITLE
fabric:query_view/4 fails if number of docs in the DB is more than 60

### ISSUE THREAD

**Issue by AlexanderKaraberov** [apache/couchdb#1142]
## Expected Behaviour
Currently I'm implementing new view filter functionality using CouchDB 2.2 (master branch) codebase. Namely I want to filter `_changes` feed using additional query params like this:
`_changes?include_docs=true&filter=_view&view=index/type&key="d"'`, `start_key_docid` and so on. 
I managed to implement a working solution. I parse query string to `#mrargs` record (I reuse `fitler_args` field) and then pass this to this function (I wrote an additional case for a 5-tuple).
```erlang
filter(Db, DocInfo, {FilterType, Style, DDoc, VName, FilterArgs})
        when FilterType == view; FilterType == fast_view
```
And then I use `query_view/4` from `fabric.erl` proxy layer in order to execute a view and receive this kind of response `{row,[{id,<<"doc2">>},{key,<<"d">>}]}]`  and then I implemented something similar to this: `{ok, Passes} = couch_query_servers:filter_view(DDoc, VName, Docs),` functionality.
I would expect this to work on any number of docs (in my case an I have in average ~50000 docs and up to 200000 docs).

## Current Behaviour
My implementation works flawlessly with very small test databases (1-55 docs). I can apply additional filters (`type`, `keys` and so on) as a query param and all my tests succeed but when I have more than 50-60 docs this function fails. I have `results [ ` printed in logs and then just an empty space. I receive no timeout as well. Just in case I also checked in `fabric_view_map:go` that timeout value for erlang:send_after is fairly large:
```erlang
case rexi_utils:recv(Workers, #shard.ref, fun handle_message/3,
        State, infinity, 1000 * 60 * 60) of
``` 
Currently for the sake of testing I'm using only map views but i see the same `rexi_util_recv` call in `fabric_view_reduce:go2` as well.

## Possible Solution
It looks like Erlang `gen_server`'s `handle_call` doesn't have enough time to receive a response chunk and just infinitely calls `changes_callback` with no result but I might be wrong. Probably that's because of clustered configuration (I have a default 3-node setup) and now `query_view` takes additional time to query all nodes but timeout for `erlang:send_after` wasn't increased.

## Steps to Reproduce (for bugs)
You can very easily reproduce this. 
1. Just insert this call to query_view in `couch_changes:filter` with filter type == `view`:
```erlang
[_, DName] = binary:split(element(2, DDoc), <<"/">>),
 QArgs = #mrargs{},
 {ok, VResp} = fabric:query_view(<<db1>>, DName, VName, QArgs),
```
2. Test with small db (<50 docs): create a simple ddoc `index`, make query:
`curl -u admin:rm http://127.0.0.1:5983/db1/_design/index/_view/type` - everything works
2 Add more docs (e.g. 90), make the same query and observe that `_changes` feed stuck.

## Context
As I said in the beginning I need this functionality to work because I want to implement my additional logic for view filtering.
P.S Yes I know that I may use mango selectors for this (in fact i already do this):
http://docs.couchdb.org/en/2.1.1/api/database/changes.html#selector
But I want to know query_view and hence my implementation doesn't work on a fairly small DB.

## Your Environment
CouchDB 2.2 (fork from master). But I already tested this with 2.1 release version and behaviour is completely the same unfortunately.

Any help from CouchDB core devs will be much appreciated!

**Comment by davisp** [comment]
First guess is that you're mixing up the rexi message handling. The rexi_util:recv function isn't written to handle multiple nested calls and will end up chewing through and mixing up messages from RPC workers. A quick test is to use something of this nature to replace you fabric:query_view call:

```erlang
do_query(DbName, VName, QArgs) ->
    {_, Ref} = spawn_monitor(fun() ->
        try
            Resp = fabric:query_view(<<"db1">>, DName, VName, QArgs)
            exit({xkcd_ok, Resp})
        catch T:R ->
            S = erlang:get_stacktrace(),
            exit({xkcd_error, T, R, S})
    end),
    receive
        {'DOWN', Ref, _, _, {xkcd_ok, Resp}} ->
            Resp
        {'DOWN', Ref, _, _, {xkcd_error, T, R, S}} ->
            erlang:raise(T, R, S)
    end.
```

**Comment by AlexanderKaraberov** [comment]
Hi @davisp , 
Thanks for the explanation! Regarding my use case (namely fine-grained filtering in `_changes` feed) I'm planning to implement an alternative implementation to avoid misuse of the ` rexi_util:recv`. Basically I want to call `fabric:query_view` only once in `chttpd_db:handle_changes_req` then store a response in the `ETS` table and then use it like this to do the actual filtering instead of `couch_query_servers:filter_view()`:
```erlang 
    Passes = case couch_util:get_value(row, lists:first(ets:lookup(Resp, ViewResp)), undefined) of
            undefined ->
              lists:map(fun (_) -> false end, Docs);
            Object ->
              lists:map(fun (_) -> true end, Docs)
            end,
    filter_revs(Passes, Docs);
```
What is your opinion about this approach? Or maybe you may give a hint on how to implement this in  a better way?

**Comment by AlexanderKaraberov** [comment]
UPD: I reimplemented `_changes` feed filtering properly without using `fabric:query_view`. My implementation is based on a `fast_view` optimisation for view-filtered `_changes` feed. I also managed to implement `fast_view` properly and made it cluster-aware by exposing it `fabric`. 
I'm going to submit a PR with this functionality + working `fast_view` soon (after I polish code and write tests).
So feel free to close this issue it it is out of scope and is caused by abuse of `rexi`/`fabric`.

**Comment by eiri** [comment]
Hi @AlexanderKaraberov - I heard that view changes implementation is _conceptually_ broken in mrview implementation, it appears to work in the simple cases, but quickly goes south once keys in views starting to be removed and then reappear. I know at least one corner case where seq btree in mrview is not working correctly. What I'm trying to say is that it's not simply a question of wiring view changes through fabric and removing `fast_view` shortcut, there are a good chance you are wasting your time on a broken feature.

**Comment by AlexanderKaraberov** [comment]
Hi @eiri 

Could you please describe this

> one corner case where seq btree in mrview is not working correctly

? 
I mean we tested this implementation using some basic test cases (mostly replication to mobile databases using Couchbase Lite client and filtering using Apache Bench for benchmarking) and everything seems to work fine. 
Just in case our implementation is based on `couch_mrview:count_view_changes_since()` +               couch_mrview:view_changes_since()`.

**Comment by eiri** [comment]
The issue with seq btree described here #592. In nutshell the problem is that unlike db's btree keys that always strings, views keys in seq btree could be anything, so sorting could yield unexpected results.

For the keys re-entering issues, I haven't ran into that myself, but views changes been discussed recently on dev channel and one of the senior developers mentioned that in scenario when the same keys appear in a view, then removed with docs updated or deleted and then reappearing again with new updates the view changes breaks pretty quickly. The general consensus was that we'd want to remove current implementation rather than trying to fix it and then redo it anew. Making a changes feed work properly is not a very trivial task as discussed on this thread #831

**Comment by AlexanderKaraberov** [comment]
@eiri Thanks for information.
You said `then redo it anew`. Is this somewhere in the roadmap? If yes i can try to help with this.
I'm asking for the reason that this functionality is crucial for us since otherwise replication to mobile clients will be broken in our case and also `_changes` without `fast_view` is super slow and unusable even for medium scale production use case unfortunately. The same is true for `mango` selectors workaround which we implemented before exposing `fast_view` to `fabric_rpc`.
Also as I understood sending a PR with `couch_mrview:count_view_changes_since()`/`couch_mrview:view_changes_since()` in `fabric_rpc` is useless because it relies on a broken `mrview` implementation?

**Comment by AlexanderKaraberov** [comment]
@eiri Also I'm not sure I totally understood you. I mean if the problem you mentioned exists than the same issue exists with normal views because views are sorted first by key.

**Comment by AlexanderKaraberov** [comment]
Hey @eiri  just want to clarify a few last things in order to sort everything out. 
As I understood by "custom sorter" you meant [view collation](http://docs.couchdb.org/en/latest/ddocs/views/collation.html#views-collation) and  seq btree/`KeyBySeqBtree` works the same as [raw collation](http://docs.couchdb.org/en/latest/ddocs/views/collation.html#raw-collation)? 
Yet I still don't understand how sorting is related to constructing output for `_changes` feed since for that we keep a separate index order by `seq` of the database.

**Comment by AlexanderKaraberov** [comment]
Also as I understood from [this issue](https://github.com/apache/couchdb/issues/592) I can fix this by providing proper collation when `kseq` is constructed. And as you said: 

> We need to specify custom less function on seq_btree creation and normalize keys in it.

**Comment by eiri** [comment]
The seq btree _is_ that separate index and its keys [are tuples of update seq and view key](https://github.com/apache/couchdb/blob/master/src/couch_mrview/src/couch_mrview_updater.erl#L104). So if your keys are strings, then you can pull changes since given update seq, because `{<<"4-g1AAAA">>, <<>>} < {<<"4-g1AAAA">>, <<"abc">>}` is `true`, but when your keys are, say, integers, you can't, since `{<<"4-g1AAAA">>, <<>>} < {<<"4-g1AAAA">>, 10}` is `false`. I hope that makes it clearer.

Yes, as I said, it's easy to remediate by setting custom collation on `seq` btree (note, that `kseq` is _different_ index and its collation set the same as view's), but it doesn't fix a deeper issue with the whole implementation.

**Comment by eiri** [comment]
> So I'm curious is it possible to re-implement Seq index a little bit...

Probably, maybe?.. I don't know, to be honest, I haven't thought too hard about it. Setting custom `less` on `seq_btree` feels like a less hacky approach to me and I'd simply serialize key to binary to take a randomness of hashing out of the equation.

I presume the keys were used on seq btree because a view can emit multiple keys per db update. Which, I imagine, would give a duplicated entires in `_changed` feed when filtering used, no idea how replicator suppose to handle that.

Honestly I'd be very wary of using a known broken feature in a production even if it works in a small subset of the tests and talking from experience migrating production views from fork to upstream, especially when file format changed, is a _huge_ PITA. But that's your call, obviously, I don't see this feature to get into active development too soon.

**Comment by AlexanderKaraberov** [comment]
> Honestly I'd be very wary of using a known broken feature in a production

Yes I know but do I really have a choice? I can't think about any other alternatives to `fast_view`. Before this we tried to use mango selectors with some hackery on the `nginx` side to make `Couchbase Lite` replication work, but unfortunately selectors also are too slow and have [some other issues](https://github.com/apache/couchdb/issues/1151). Simply avoiding using `seq` and `kseq` indexes proved to be outright slow and completely unusable in production scale. I'm not sure there is a forth alternative currently.

**Comment by eiri** [comment]
I don't know much about Couchbase Lite, but in Couch there are also native erlang views, though they are their own can of worms around security afaik. However I see you point, just thought it'd be fair to give you a warning.
